src/main/java/de/tum/in/www1/exerciseapp/service/CourseService.java

```diff
@@ -149,29 +151,39 @@ public void delete(Long id) {
      * Find the best Result in a Participation
      *
      * @param participation the participation you want the best result from
+     * @param hasDueDate if the participation has a duedate take last result before the due date if not take the overall last result
      * @return the best result a student had within the time of the exercise
      */
     @Transactional(readOnly = true)
-    public Result bestResultScoreInParticipation(Participation participation) {
+    public Result bestResultScoreInParticipation(Participation participation, boolean hasDueDate) {
         Set<Result> results = participation.getResults();
 
-        Result bestResult = null;
+        Result bestResult;
+
+        //edge case of no result submitted to a participation
+        if(results.size() <= 0){
+            bestResult = new Result();
+            bestResult.setScore((long) 0);
+            bestResult.setParticipation(participation);
+            return bestResult;
+        }
+
+        bestResult = results.iterator().next();
         for (Result result : results) {
-            if (bestResult == null) {
-                bestResult = new Result();
-                bestResult.setScore(result.getScore());
-            } else if (bestResult.getScore() != null
-                && result.getScore() != null //score can be stored as null in the DB
-                && bestResult.getScore() < result.getScore()
-                && participation.getExercise().getDueDate().isBefore(result.getCompletionDate())//only use results which are completed before the dueDate
-                ) {
-                bestResult.setScore(result.getScore());
+            // the participation has a due date take last result before the due date
+            if(hasDueDate
+                && bestResult.getCompletionDate().isBefore(result.getCompletionDate())
+                && result.getCompletionDate().isBefore(participation.getExercise().getDueDate())
+                ){
+                bestResult = result;
+            }else if( bestResult.getCompletionDate().isBefore(result.getCompletionDate()) // take the overall last submitted result
+                ){
+                bestResult=result;
             }
```

Comment:
should now chose the result accordingly to:

Case 1: there is a due date:
first determine the latest result within the due date (i.e. the one that is closest to the due date, but still before it), and simply use this result
Case 2: there is no due date:
simply take the result with the latest completion date